agent/provider/src/market/provider_market.rs

```diff
@@ -594,6 +658,55 @@ impl Handler<AgreementFinalized> for ProviderMarket {
                 error,
             );
         }
+
+        let myself = ctx.address();
+        let market_api = self.market_api.clone();
+        let subscriptions = self.offer_subscriptions.clone();
+        let negotiations = std::mem::replace(&mut self.postponed_negotiations, HashMap::new());
+
+        let fut = async move {
+            let dt = Duration::seconds(POSTPONED_NEGOTIATION_TIMEOUT_S);
+            let iter = negotiations.into_iter().choose(&mut rand::thread_rng());
+
+            for (subscription_id, (proposal, date_time)) in iter {
+                let now = chrono::Utc::now().naive_utc();
+                if now - date_time > dt {
+                    log::info!(
+                        "Postponed negotiation for subscription [{}] timed out.",
+                        subscription_id
+                    );
+                    continue;
+                }
+
+                let subscription = match subscriptions.get(&subscription_id) {
+                    Some(sub) => sub.clone(),
+                    None => {
+                        log::info!(
+                            "Ignoring postponed negotiation: subscription [{}] no longer exists.",
+                            subscription_id
+                        );
+                        continue;
+                    }
+                };
+
```

Comment:
```suggestion
                    log::info!(
                        "Taking again postponed negotiation of Proposal [{}] for subscription [{}].",
                        proposal.proposal_id()?,
                        subscription_id
                    );
```

Reply:
`proposal.proposal_id()?` - `?` isn't allowed here

Reply:
your change is even better than my proposition